File: scrape.js
const axios = require("axios");
const cheerio = require("cheerio");
const fs = require("fs");

const base_url = "https://fallout.fandom.com";

async function getQuotes(quotes, url) {
	// tags of interest
	// ================
	// "div.mediaContainer" - holds the audio file player
	//
	// "li" - in "Notable Quotes" section, contains quote and "div.np-quote-snd-player"
	// "div.np-quote-snd-player" contains href to audio file.
	//
	// "span.va-quotation-text" - top-of-page quote
	//
	// "div.lightbox-caption" - contains a quote and a "div.mediaContainer"
	//
	// "div.np-collapsible" - contains a "ul" of quotes

	const { data } = await axios.get(base_url.concat(url)).catch( error => {
		console.error(`Error loading page: ${url}`);
		return;
	});
	const $ = cheerio.load(data);

	const page_title = url.split("/").pop();
	quotes["pages"][page_title] = [];
	const page = {"page_title": page_title, "quotes": []};

	var quote_section = $("h2:contains('Notable quotes')")
		.nextUntil("h2")
		.find("li");
	quote_section.find("a,span").text("");
	for (let i = 0; i < quote_section.length; i++) {
		let li = quote_section.eq(i);
		let quote_text = li.text();
		let audio_url = li.find("source").attr("src");
		quotes["pages"][page_title]["quotes"].push {
			"text": quote_text,
			"audio": audio_url,
		});
		console.log(quote_text);
	}

	console.log(`Got ${page["quotes"].length} quotes from ${url}`);
}

async function crawl(quotes, url) {

	const page_title = url.split("/").pop().toLowerCase();
	if (page_title.startsWith("file:")) {
		return;
	} else if (page_title.startsWith("category:") === false) {
		return getQuotes(quotes, url);
	}

	const { data } = await axios.get(base_url.concat(url)).catch(error => {
		console.error(`Error loading page: ${url}`);
		return;
	});

	const $ = cheerio.load(data);

	var members = $("div.category-page__members").find(".category-page__member-link");
	for (let i = 0; i < members.length; i++) {
		crawl(quotes, members.eq(i).attr("href"));
	}
}

function getCharQuotes() {
	const start_url ="/wiki/Category:Characters_by_game";
	const quotes = {"pages": {}};
	crawl(quotes, start_url);
	return quotes;
	console.log("Writing to quotes.json...");
	fs.writeFile("quotes.json", JSON.stringify(quotes), "utf8", (err) => {
		if (err) {
			console.log(`Error writing file: ${err}`);
		} else {
			console.log("Quote database successfully saved!");
		}
	});
}

getCharQuotes();

File: scrape.py
import sys
from bs4 import BeautifulSoup as Soup
from urllib import request
import json

base_url = "https://fallout.fandom.com"

def getQuotes(href):
    # tags of interest
    # ================
    # "div.mediaContainer" - holds the audio file plyaer
    #
    # "li" - in "Notable Quotes" section, contains quote and "div.np-quote-snd-player"
    # "div.np-quote-snd-player" contains href to audio file.
    #
    # "span.va-quotation-text" - top-of-page quote
    #
    # "div.lightbox-caption" - contains a quote and a "div.mediaContainer"
    #
    # "div.np-collapsible" - contains a "ul" of quotes

    url = base_url + href

    print("Getting quotes from page:", href)

    with request.urlopen(url) as response:
        quotes = []
        html = response.read()
        soup = Soup(html, features="html5lib")

        page_title = url.split("/").pop()

        el = soup.select_one("h2:-soup-contains('Notable quotes')")
        if el:
            el = el.next_element
            while el.name != "h2":
                if el.name == "li":

                    s = ".np-quote-snd-player-title a, np-quote-snd-player-title span"
                    for e in el.select(s):
                        e.string.replace_with("")

                    quote_text = el.string
                    audio_url = el.find("source")
                    if audio_url:
                        audio_url = audio_url["src"]
                    else:
                        autio_url = ""

                    new_quote = {
                        "quote_text": quote_text,
                        "audio_url": audio_url,
                        "page_title": page_title
                    }
                    quotes.append(new_quote)
                    print("Got quote:", new_quote)

                elif el.name == "h2":
                    break

                el = el.next_element

        return quotes

def crawl(href):
    page_title = href.split("/").pop().lower()
    if (page_title.startswith("file:")):
        return []
    elif (page_title.startswith("category:") == False):
        return getQuotes(href)

    quotes = []
    url = base_url + href
    with request.urlopen(url) as response:
        html = response.read()
        soup = Soup(html, features="html5lib")

        members = soup.select("div.category-page__members .category-page__member-link")
        if members:
            for m in members:
                quotes += crawl(m["href"])

    return quotes

if __name__ == "__main__":
    href = '/wiki/Category:Characters_by_game'
    quotes = crawl(href)
    quotes += getQuotes(url)
    with open("quotes.json", "w") as quoteFile:
        json.dump(quotes, quoteFile)
